src/common/cuda_pinned_allocator.h

```diff
@@ -1,93 +1,114 @@
-/*!
- * Copyright 2022 by XGBoost Contributors
- * \file common.h
- * \brief cuda pinned allocator for usage with thrust containers
+/**
+ * Copyright 2022-2024, XGBoost Contributors
+ *
+ * @brief cuda pinned allocator for usage with thrust containers
  */
 
 #pragma once
 
+#include <cuda_runtime.h>
+
 #include <cstddef>
 #include <limits>
 
 #include "common.h"
 
-namespace xgboost {
-namespace common {
-namespace cuda {
-
+namespace xgboost::common::cuda_impl {
 // \p pinned_allocator is a CUDA-specific host memory allocator
 //  that employs \c cudaMallocHost for allocation.
 //
 // This implementation is ported from the experimental/pinned_allocator
 // that Thrust used to provide.
 //
 //  \see https://en.cppreference.com/w/cpp/memory/allocator
+
 template <typename T>
-class pinned_allocator;
+struct PinnedAllocPolicy {
+  using pointer = T*;              // NOLINT: The type returned by address() / allocate()
+  using const_pointer = const T*;  // NOLINT: The type returned by address()
+  using size_type = std::size_t;   // NOLINT: The type used for the size of the allocation
+  using value_type = T;            // NOLINT: The type of the elements in the allocator
+
+  size_type max_size() const {  // NOLINT
+    return (std::numeric_limits<size_type>::max)() / sizeof(value_type);
```

Comment:
Shouldn't it just be:
```cpp
return std::numeric_limits<size_type>::max() / sizeof(value_type);
```

Reply:
Yeah, that's from the old code, haven't done the cleanup yet, will change it.